File: gaiag/baseline/Guardthreetopon.cc
// Dezyne --- Dezyne command line tools
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2015 Rutger van Beusekom <rutger.van.beusekom@verum.com>
// Copyright © 2015 Paul Hoogendijk <paul.hoogendijk@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include "Guardthreetopon.hh"

#include "locator.hh"
#include "runtime.hh"

#include <iostream>

namespace dezyne
{
  Guardthreetopon::Guardthreetopon(const locator& dezyne_locator)
  : dzn_meta{"","Guardthreetopon",reinterpret_cast<const component*>(this),0,{},{[this]{i.check_bindings();},[this]{r.check_bindings();}}}
  , dzn_rt(dezyne_locator.get<runtime>())
  , b(false)
  , i({{"i",this},{"",0}})
  , r({{"",0},{"r",this}})
  {
    dzn_rt.performs_flush(this) = true; 
    i.in.e = [&] () {
      call_in(this, [this] {i_e();}, std::make_tuple(&i, "e", "return"));
    };
    i.in.t = [&] () {
      call_in(this, [this] {i_t();}, std::make_tuple(&i, "t", "return"));
    };
    i.in.s = [&] () {
      call_in(this, [this] {i_s();}, std::make_tuple(&i, "s", "return"));
    };
    r.out.a = [&] () {
      call_out(this, [this] {r_a();}, std::make_tuple(&r, "a", "return"));
    };

  }

  void Guardthreetopon::i_e()
  {
    if (true and b)
    {
      i.out.a();
    }
    else if (true and not (b))
    {
      bool c = true;
      if (c)
      i.out.a();
    }
  }

  void Guardthreetopon::i_t()
  {
    if (b)
    i.out.a();
    else if (not (b))
    i.out.a();
  }

  void Guardthreetopon::i_s()
  {
    i.out.a();
  }

  void Guardthreetopon::r_a()
  {
    {
    }
  }


  void Guardthreetopon::check_bindings() const
  {
    dezyne::check_bindings(reinterpret_cast<const dezyne::component*>(this));
  }
  void Guardthreetopon::dump_tree() const
  {
    dezyne::dump_tree(reinterpret_cast<const dezyne::component*>(this));
  }
}

File: gaiag/baseline/Guardthreetopon.java
// Dezyne --- Dezyne command line tools
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

class Guardthreetopon extends Component {

  Boolean b;

  IGuardthreetopon i;
  RGuardthreetopon r;

  public Guardthreetopon(Runtime runtime) {this(runtime, "");};

  public Guardthreetopon(Runtime runtime, String name) {this(runtime, name, null);};

  public Guardthreetopon(Runtime runtime, String name, SystemComponent parent) {
    super(runtime, name, parent);
    this.flushes = true;
    b = false;
    i = new IGuardthreetopon();
    i.in.name = "i";
    i.in.self = this;
    r = new RGuardthreetopon();
    r.out.name = "r";
    r.out.self = this;
    i.in.e = new Action() {public void action() {Runtime.callIn(Guardthreetopon.this, new Action() {public void action() {i_e();}}, new Meta(Guardthreetopon.this.i, "e"));};};

    i.in.t = new Action() {public void action() {Runtime.callIn(Guardthreetopon.this, new Action() {public void action() {i_t();}}, new Meta(Guardthreetopon.this.i, "t"));};};

    i.in.s = new Action() {public void action() {Runtime.callIn(Guardthreetopon.this, new Action() {public void action() {i_s();}}, new Meta(Guardthreetopon.this.i, "s"));};};

    r.out.a = new Action() {public void action() {Runtime.callOut(Guardthreetopon.this, new Action() {public void action() {r_a();}}, new Meta(Guardthreetopon.this.r, "a"));};};

  };
  public void i_e() {
    if (true && b) {
      i.out.a.action();
    }
    else if (true && ! (b)) {
      V<Boolean> c = new V <Boolean>(true);
      if (c.v) i.out.a.action();
    }
  };

  public void i_t() {
    if (b) i.out.a.action();
    else if (! (b)) i.out.a.action();
  };

  public void i_s() {
    i.out.a.action();
  };

  public void r_a() {
    { }
  };

}
